saturn_client/core.py

```diff
@@ -143,6 +169,172 @@ def create_project(
+
+        project_url = urljoin(self.url, f"api/projects/{project_id}")
+        response = requests.patch(
+            project_url,
+            data=json.dumps(project_config),
+            headers=self.settings.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as err:
+            raise HTTPError(
+                response.status_code, response.json()["message"] + _maybe_name(project_id)
+            ) from err
+        project = response.json()
+
+        if not (project["jupyter_server_id"] and update_jupyter_server):
+            return project
+
+        jupyter_config = {
+            "image": image_uri,
+            "start_script": start_script,
+            "environment_variables": environment_variables,
+            "working_dir": working_dir,
+            **ws.jupyter_server_settings,
+        }
+        # only send kwargs that are explicitly set by user
+        jupyter_config = {k: v for k, v in jupyter_config.items() if v is not None}
+        if len(jupyter_config) == 0:
+            return project
+
+        self.stop_jupyter_server(project["jupyter_server_id"])
+        jupyter_url = urljoin(self.url, f"api/jupyter_servers/{project['jupyter_server_id']}")
+        response = requests.patch(
+            jupyter_url,
+            data=json.dumps(jupyter_config),
+            headers=self.settings.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as err:
+            raise HTTPError(response.status_code, response.json()["message"]) from err
+        return project
+
+    def get_jupyter_server(self, jupyter_server_id) -> Dict[str, Any]:
+        """Get a particular jupyter server"""
+        url = urljoin(self.url, f"api/jupyter_servers/{jupyter_server_id}")
+        response = requests.get(
+            url,
+            headers=self.settings.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as err:
+            raise HTTPError(
+                response.status_code, response.json()["message"] + _maybe_name(jupyter_server_id)
+            ) from err
+        return response.json()
+
+    def stop_jupyter_server(self, jupyter_server_id):
+        """Stop a particular jupyter server"""
+        url = urljoin(self.url, f"api/jupyter_servers/{jupyter_server_id}/stop")
+        response = requests.post(
+            url,
+            data=json.dumps({}),
+            headers=self.settings.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as err:
+            raise HTTPError(
+                response.status_code, response.json()["message"] + _maybe_name(jupyter_server_id)
+            ) from err
+
+    def start_jupyter_server(self, jupyter_server_id):
+        """Start a particular jupyter server"""
+        url = urljoin(self.url, f"api/jupyter_servers/{jupyter_server_id}/start")
```

Comment:
If we're exposing a `start()` method, I think we have to expose either `/api/jupyter_servers/{id}/runtimesummary` or a simpler `/api/jupyter_servers/{id}/status` method. Since this method will return successfully as soon as Saturn says "ok I know what you want me to do and this is real Jupyter", client code needs some other way to block until the Jupyter has been started. Like a method that accomplishes what I've done with raw HTTP requests in some integration tests:

```python
logger.info("Waiting for Jupyter to start...")
retries_so_far = 0
while retries_so_far < MAX_RETRIES:
    retries_so_far += 1
    logger.info(f"Checking jupyter status ({retries_so_far}/{MAX_RETRIES})")
    result = requests.get(
        url=f"{SATURN_HOST}/api/jupyter_servers/{jupyter_id}/runtimesummary",
        headers=SATURN_AUTH_HEADERS,
    )
    if result.status_code == 200:
        if result.json()["status"] == "running":
            logger.info("Jupyter started successfully")
            break
        else:
            time.sleep(SLEEP_SECONDS_BETWEEN_RETRIES)
    else:
        result.raise_for_status()
```

My vote is for a method that is like `wait_for_jupyter()`, that mimics the behavior of Dask's `wait_for_workers()` and handles the polling internally. That way, we don't have to explicitly expose the status strings to customer code.

This could be done in a later PR, but it shouldn't be too much later. Without it, I expect that people using `start_jupyter()` will follow it with a `time.sleep(20)` or something, which is inefficient and not fun.

Reply:
Yes :+1: I have a note about this behavior in the README, but I should put it in the docstring as well. This is definitely something that shouldn't wait too long.